### PR TITLE
Context propagation - OTEP-66

### PR DESCRIPTION
**Pull request by MikeGoldsmith**
Part of #461.

This is the first step in separating context propagation in accordance with the recently merged [OTEP-66](https://github.com/open-telemetry/oteps/blob/master/text/0066-separate-context-propagation.md).

This PR has the following changes:
- renames DistributedContext > CorrelationContext, DistributedContextEntry > CorrelationContextEntry & DistributedContextEntryMetadata > CorrelationContextEntryMetda
- add new DistributedContext that contains both a CorrelationContext and SpanContext

The recently merged #444 reworked DistributedContext to use a builder pattern, which I'm currently merging into my local branch so I've got it as draft.

### CODE REVIEW

**Comment by lmolkova** [comment]
@tarekgh can you please have a look?

File: src/OpenTelemetry.Api/Correlations/CorrelationContextEntry.cs [hunk 1]
```diff
@@ -16,35 +16,35 @@
 
 using OpenTelemetry.Internal;
 
-namespace OpenTelemetry.Context
```
Comment (tarekgh) [inline src/OpenTelemetry.Api/Correlations/CorrelationContextEntry.cs]: would it make sense to keep the namespace Context? it is still the right thing. I am imagining in the future we can have one more type of context objects and would make sense to keep all of them under one namespace.

File: src/OpenTelemetry.Api/Correlations/CorrelationContextEntry.cs [hunk 2]
```diff
@@ -16,35 +16,35 @@
...
         /// <summary>
-        /// Initializes a new instance of the <see cref="DistributedContextEntry"/> struct with the key and value.
+        /// Initializes a new instance of the <see cref="CorrelationContextEntry"/> struct with the key and value.
         /// </summary>
         /// <param name="key">Key name for the entry.</param>
         /// <param name="value">Value associated with the key name.</param>
-        public DistributedContextEntry(string key, string value)
-            : this(key, value, EntryMetadata.NoPropagationEntry)
+        public CorrelationContextEntry(string key, string value)
+            : this(key, value, CorrelationEntryMetadata.NoPropagationEntry)
```
Comment (tarekgh) [inline src/OpenTelemetry.Api/Correlations/CorrelationContextEntry.cs]: would it make sense to keep it as EntryMetadata? I guess it is possible can be used in other places outside the correlation context if we need to. what do you think?

**Comment by tarekgh** [comment]
I looked at OTEP-66 and I have a question that may help me understand the change. 

- OTEP-66 making the contexts in its layer to allow using it without depending on tracing/metrics or any other layers. The question I have, why we need to combine all context objects (e.g. SpanContext and CorrelationContext) into one context object? 
    - Propagator interfaces can still deal with different context objects and not necessary to deal with one big context object. it has to deal with every context object anyway.
    - I think we can have separate context objects in the same layer as OTEP-66 suggested without combining them.
    - The container of context objects has to deal with multiple context objects while maybe in the most common scenarios users care about only one context type.
    - Combining the context objects into one object can create some confusion. For example, usually, SpanContext is currently tied to the span which is tied to Activity. The activity has its own scoping (through asynclocal). Now having SpanContext inside global context container object which has its own scoping (not connected to Activity scoping). Is this a scenario we need to support? Does SpanContext desire in non-span scenarios?
    - in the future, if we got any new context scenario which needs a new context class, how the contexts container will look like? 

I admit I am not seeing the big picture but the answers to these questions can help here.

File: src/OpenTelemetry.Api/Correlations/CorrelationContextEntry.cs (hunk 1 above)
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Api/Correlations/CorrelationContextEntry.cs]: I kept the Context namespace.

File: src/OpenTelemetry.Api/Correlations/CorrelationContextEntry.cs (hunk 2 above)
Comment (MikeGoldsmith) [reply src/OpenTelemetry.Api/Correlations/CorrelationContextEntry.cs]: I left this as EmtryMetadata.

**Comment by MikeGoldsmith** [comment]
Finally got round to picking this up and finishing it up.

The steps I took are:
- rename DistributedContext and all references to it, to CorrelationContext
- add new DistributedContext as wrapper that contains a ref to a CorrelationContext
- add DistributedContextBuilder
- update DistributedContextCarrier to work with DistributedContext

I'm not 100% sure on the new DistributedContextBuilder, so feedback on that would be appreciated.

**Comment by cijothomas** [comment]
@MikeGoldsmith Can you resolve conflicts and get changes from master. Else its difficult to review as there are a lot of unrelated changes.

**Comment by MikeGoldsmith** [comment]
Sure, will do - looks like it got out of date pretty quickly!

**Comment by MikeGoldsmith** [comment]
This is ready for review after resolving all the conflicts in merging master into my branch.

@open-telemetry/dotnet-approvers

File: src/OpenTelemetry.Api/Context/CorrelationContext.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    public readonly struct CorrelationContext : IEquatable<CorrelationContext>
+    {
+        private static readonly List<CorrelationContextEntry> EmptyList = new List<CorrelationContextEntry>();
+        private readonly List<CorrelationContextEntry> entries;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CorrelationContext"/> struct.
+        /// </summary>
+        /// <param name="entries">Entries for correlation context.</param>
+        internal CorrelationContext(List<CorrelationContextEntry> entries)
```
Comment (reyang) [inline src/OpenTelemetry.Api/Context/CorrelationContext.cs:35]: Curious, is there a limitation on the number of entries?

File: src/OpenTelemetry.Api/Context/CorrelationContext.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        /// <summary>
+        /// Gets all the <see cref="CorrelationContextEntry"/> in this <see cref="CorrelationContext"/>.
+        /// </summary>
+        public IEnumerable<CorrelationContextEntry> Entries => this.entries;
+
+        /// <summary>
+        /// Gets the <see cref="CorrelationContextEntry"/> with the specified name.
+        /// </summary>
+        /// <param name="key">Name of the <see cref="CorrelationContextEntry"/> to get.</param>
+        /// <returns>The <see cref="CorrelationContextEntry"/> with the specified name. If not found - null.</returns>
```
Comment (cijothomas) [inline src/OpenTelemetry.Api/Context/CorrelationContext.cs:54]: this method is returning the string value, not `CorrelationContextEntry`

File: src/OpenTelemetry.Api/Context/CorrelationContext.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+using System.Linq;
+
+namespace OpenTelemetry.Context
+{
+    /// <summary>
+    /// Correlation context.
+    /// </summary>
+    public readonly struct CorrelationContext : IEquatable<CorrelationContext>
+    {
+        private static readonly List<CorrelationContextEntry> EmptyList = new List<CorrelationContextEntry>();
```
Comment (tarekgh) [inline src/OpenTelemetry.Api/Context/CorrelationContext.cs:28]: >private static readonly List<CorrelationContextEntry> EmptyList = new List<CorrelationContextEntry>(); [](start = 8, length = 102)

nit: I am seeing this is used only in CorrelationContext Empty { get; } so I am not sure if it worth to declare it. 

you can just have.

public static CorrelationContext Empty { get; } = new CorrelationContext(new  List<CorrelationContextEntry>(););

it is ok to keep it if you think we are going to use it for other purpose.

**Comment by tarekgh** [comment]
>             return EmptyByteArray;

nit: maybe better to use 

[https://docs.microsoft.com/en-us/dotnet/api/system.array.empty?view=netframework-4.8](https://docs.microsoft.com/en-us/dotnet/api/system.array.empty?view=netframework-4.8)

---
Refers to: src/OpenTelemetry/Context/NoopDistributedContextBinarySerializer.cs:34 in 4a293aa. [](commit_id = 4a293aa204353d10779c7c6d5d869024a5e04b5c, deletion_comment = False)

File: src/OpenTelemetry.Api/Context/CorrelationContext.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        /// Gets the <see cref="CorrelationContextEntry"/> with the specified name.
+        /// </summary>
+        /// <param name="key">Name of the <see cref="CorrelationContextEntry"/> to get.</param>
+        /// <returns>The <see cref="CorrelationContextEntry"/> with the specified name. If not found - null.</returns>
+        public string GetEntryValue(string key) => this.entries.LastOrDefault(x => x.Key == key).Value;
+
+        /// <inheritdoc/>
+        public bool Equals(CorrelationContext other)
+        {
+            if (this.entries.Count() != other.entries.Count())
```
Comment (tarekgh) [inline src/OpenTelemetry.Api/Context/CorrelationContext.cs:60]: >entries [](start = 21, length = 7)

I think we need to check on null entries

**Comment by tarekgh** [comment]
@MikeGoldsmith thanks for getting ready. I added a few comments but other than that LGTM. 

Quick question, what is the benefit of having DistributedContext? so far it looks like it is just a wrapper to CorrelationContext. Are we going to add more to DistributedContext later or in the future?

File: src/OpenTelemetry.Api/Context/CorrelationContext.cs (hunk 4 above)
Comment (bruno-garcia) [inline src/OpenTelemetry.Api/Context/CorrelationContext.cs:54]:
```suggestion
        /// <returns>The <see cref="string"/> with the specified name. If not found - null.</returns>
```

File: src/OpenTelemetry.Api/Context/DistributedContextBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+            new DistributedContext(new CorrelationContextBuilder(inheritCurrentContext: false).Add(entries).Build());
+
+        /// <summary>
+        /// Configures correlations to be used with the context.
+        /// </summary>
+        /// <param name="configureCorrelations">An <see cref="Action{CorrelationContextBuilder}"/> used to configure correlations.</param>
+        /// <returns>The current <see cref="DistributedContextBuilder"/> instance.</returns>
+        public DistributedContextBuilder Correlations(Action<CorrelationContextBuilder> configureCorrelations)
+        {
+            configureCorrelations(this.correlationContextBuilder);
```
Comment (bruno-garcia) [inline src/OpenTelemetry.Api/Context/DistributedContextBuilder.cs:72]: Should we deal with null here?

```suggestion
            configureCorrelations?.Invoke(this.correlationContextBuilder);
```

**Comment by MikeGoldsmith** [comment]
@tarekgh Yes, other context information will be added to `DistributedContext` in the future. eg MetricContext, LogContext, etc.

Thanks for the feedback @reyang @cijothomas @bruno-garcia - I'll work though it shortly.

**Comment by SergeyKanzhelev** [comment]
This seems to be a solid first step. @tarekgh please raise issues if you still has concerns

**Pull request merged**